### PR TITLE
south/ocnos: initial support for configuring interface and vlan of containerized OcNOS

### PR DESCRIPTION
**Pull request by KazuyaAnazawa**
This PR adds new agent called `south-ocnos`.
The `south-ocnos` is one of Goldstone management south daemons that control switch ASIC by configuring containerized OcNOS.
It supports the interface and vlan configuration on the platform where containerized OcNOS is running.
This implementation is contributed by IP Infusion and NTT.

### CODE REVIEW

File: src/south/ocnos/goldstone/south/ocnos/interfaces.py [hunk 1]
```diff
@@ -0,0 +1,1036 @@
...
+                                )
+                            )
+                            is_primary = False
+                        else:
+                            xpath = IPI_INTERFACE_IPV4_SECONDARY_ADDRESS_CONFIG_IP_ADDRESS.format(
+                                ifname, value
+                            )
+
+                        self.reconcile_edit_config("ip", xpath, value)
+                """
```
Comment (ishidawataru) [inline src/south/ocnos/goldstone/south/ocnos/interfaces.py:823]: ditto

File: src/south/ocnos/goldstone/south/ocnos/ocnos.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+        # for k, ms in self._models.items():
+        #    logger.debug(f"parsing {k} ..")
+        #    for m in ms:
+        #        if not m["import-only"]:
+        #            schema = m["schema"]
+        #            self.ctx.parse_module_str(schema)
+        # for m in self.ctx:
+        #    logger.info(
+        #        f"loaded module: {m}, revisions: {', '.join(str(r) for r in m.revisions())}"
+        #    )
```
Comment (ishidawataru) [inline src/south/ocnos/goldstone/south/ocnos/ocnos.py:97]: remove

File: src/south/ocnos/goldstone/south/ocnos/vlan.py [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+        )
+
+        super().stop()
+
+    def pre(self, user):
+        # TODO
+        pass
+
+    async def post(self, user):
+        # TODO
```
Comment (ishidawataru) [inline src/south/ocnos/goldstone/south/ocnos/vlan.py:236]: what is TODO here?

File: src/south/ocnos/goldstone/south/ocnos/vlan.py [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+        # Remove default network-instance
+        self._delete_default_network_instance(
+            INSTANCE_NAME_FOR_IEEE_VLAN_BRIDGE, INSTANCE_TYPE
+        )
+
+        super().stop()
+
+    def pre(self, user):
+        # TODO
+        pass
```
Comment (ishidawataru) [inline src/south/ocnos/goldstone/south/ocnos/vlan.py:233]: remove

**Review by ishidawataru — CHANGES_REQUESTED** [review]
please remove unused code

File: src/south/ocnos/goldstone/south/ocnos/interfaces.py [hunk 5]
```diff
@@ -0,0 +1,1036 @@
...
+        if self.type == "modified":
+            if self.change.prev_val == False and value == False:
+                return
+        if self.type == "created":
+            if value == False:
+                return
+
+        xpath_eth_speed = IPI_INTERFACE_ETHERNET_CONFIG_PORT_SPEED.format(self.ifname)
+
+        delete_oper = True if value == False else False
```
[the remark this answers is not in the document]
Comment (KazuyaAnazawa) [reply src/south/ocnos/goldstone/south/ocnos/interfaces.py]: Thank you for pointing it out.
Fixed it with another [similar part](https://github.com/oopt-goldstone/goldstone-mgmt/pull/71/commits/f3f32887ffe8e36dc8cd136ef930ed223e40ec72#diff-64dd4da385d6884571543f3fee097fba863d284e8a6dee1ae5f2b7ac2f03a8f6R75).

File: src/south/ocnos/goldstone/south/ocnos/vlan.py (hunk 3 above)
Comment (KazuyaAnazawa) [reply src/south/ocnos/goldstone/south/ocnos/vlan.py]: Sorry, just forgot to delete the TODO line.
Deleted it.

**Comment by KazuyaAnazawa** [comment]
@ishidawataru 
Thank you for your review. 
Reflected your comments and pushed the commit again.

File: src/south/ocnos/goldstone/south/ocnos/interfaces.py [hunk 6]
```diff
@@ -0,0 +1,917 @@
...
+            default={},
+            strip=False,
+            include_implicit_defaults=True,
+        )
+        xpath_gs_if_mode = GS_IF_MODE_CONFIG.format(self.ifname)
+        if_mode = libyang.xpath_get(running_config, xpath_gs_if_mode, None)
+        if if_mode is None:
+            if self.type == "deleted":
+                global GS_IF_MODE
+                if_mode = GS_IF_MODE.get(self.ifname)
```
Comment (ishidawataru) [inline src/south/ocnos/goldstone/south/ocnos/interfaces.py:332]: What if `interface-mode` is already configured when `south-ocnos` is started, and the configuration is handled in `reconcile()` then the `interface-mode` config is deleted? `GS_IF_MODE` doesn't have any entry, and `if_mode` will become `None` here.

Comment (ishidawataru) [reply src/south/ocnos/goldstone/south/ocnos/interfaces.py:332]: Also, `GS_IF_MODE` doesn't need to be a global variable. Please store it as a member of `InterfaceServer` and access it like `self.server.gs_if_mode` in the change handler.

**Review by ishidawataru — CHANGES_REQUESTED** [review]

File: src/south/ocnos/goldstone/south/ocnos/vlan.py (hunk 3 above)
Comment (ishidawataru) [reply src/south/ocnos/goldstone/south/ocnos/vlan.py]: please check other similar places.

https://github.com/oopt-goldstone/goldstone-mgmt/pull/71/files#diff-64dd4da385d6884571543f3fee097fba863d284e8a6dee1ae5f2b7ac2f03a8f6R87-R92

File: src/south/ocnos/goldstone/south/ocnos/interfaces.py (hunk 6 above)
Comment (KazuyaAnazawa) [reply src/south/ocnos/goldstone/south/ocnos/interfaces.py]: > What if `interface-mode` is already configured when `south-ocnos` is started, and the configuration is handled in `reconcile()` then the `interface-mode` config is deleted? `GS_IF_MODE` doesn't have any entry, and `if_mode` will become `None` here.

Current implementation does **NOT** delete `interface-mode` in `reconcile()` in that case.
When we unset `interface-mode` for an interface, we need to send the configured value (`access` or `trunk`) to OcNOS. However, we cannot obtain that value in `ChangeDeleted` callback. It seems that only xpath is obtainable like:
`event: done, changes: [ChangeDeleted(/goldstone-interfaces:interfaces/interface[name='xe1']/goldstone-vlan:switched-vlan/config/interface-mode)]`

Also, it seems that the value has already been deleted from running datastore at that time.
So we cannot get it from the datastore and the value returned [here](https://github.com/oopt-goldstone/goldstone-mgmt/pull/71/files#diff-64dd4da385d6884571543f3fee097fba863d284e8a6dee1ae5f2b7ac2f03a8f6R306-R311) is `None`.
That's why we introduced the conditional branch and global variable of `GS_IF_MODE`.

> Please store it as a member of InterfaceServer and access it like self.server.gs_if_mode in the change handler.

Thank you so much for telling me more clever way. I fixed it.

File: src/south/ocnos/goldstone/south/ocnos/vlan.py (hunk 3 above)
Comment (KazuyaAnazawa) [reply src/south/ocnos/goldstone/south/ocnos/vlan.py]: Thank you for pointing it out. I checked similar places and deleted unnecessary lines.

File: src/south/ocnos/goldstone/south/ocnos/interfaces.py [hunk 7]
```diff
@@ -0,0 +1,898 @@
...
+    "SPEED_2500M": "2.5g",
+    "SPEED_10G": "10g",
+    "SPEED_20G": "20g",
+    "SPEED_25G": "25g",
+    "SPEED_40G": "40g",
+    "SPEED_50G": "50g",
+    "SPEED_100G": "100g",
+}
+
+GOLDSTONE_INTERFACE_TOP = "/goldstone-interfaces:interfaces/interface"
```
Comment (ishidawataru) [inline src/south/ocnos/goldstone/south/ocnos/interfaces.py:23]: move these lines to `util.py`

File: src/south/ocnos/goldstone/south/ocnos/interfaces.py [hunk 8]
```diff
@@ -0,0 +1,898 @@
...
+
+
+class AdminStatusHandler(IfChangeHandler):
+    def apply(self, user):
+        if self.type in ["created", "modified"]:
+            value = self.change.value
+            if type(value) is dict:
+                return
+        else:
+            # admin-status is UP in default in OcNOS
```
Comment (ishidawataru) [inline src/south/ocnos/goldstone/south/ocnos/interfaces.py:47]: we should use the default value of Goldstone, not OcNOS.

When users delete the admin-status configuration, they expect that the default value of the Goldstone YANG model is used, which is `DOWN`.

https://github.com/oopt-goldstone/goldstone-mgmt/blob/master/yang/goldstone-interfaces.yang#L850

**Review by ishidawataru — CHANGES_REQUESTED** [review]

**Comment by ishidawataru** [comment]
LGTM. please clean up the commits.

**Comment by KazuyaAnazawa** [comment]
> LGTM. please clean up the commits.

Thank you. Cleaned up the commits.

**Review by ishidawataru — APPROVED** [review]

**Pull request merged**